### PR TITLE
[PhpUnitBridge] Replace ErrorAssert by `@expectedDeprecation`

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 3.2 |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18880 |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/7074 |

For 3.2, that's what the feat. freeze is for in this case. ping @xabbuh 

See https://github.com/symfony/symfony/pull/20255/files?w=1

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
capture du 2016-10-20 15-27-53

**Comment by fabpot** [comment]
I let @xabbuh review this PR.

One remark though: In https://thephp.cc/news/2016/02/questioning-phpunit-best-practices, Sebastian talks about the annotation and "deprecated" it as a best practice. So, I'm wondering if we don't also need to follow along and at least have the option to not use an annotation here, something like `->expectDeprecation(...)` to be consistent with PHPUnit.

**Comment by nicolas-grekas** [comment]
There is no way today to add methods to the base phpunit TestCase class (no `__call()` hook). Which means adding some `$this->expectDeprecation()` needs to be done with a trait, which then needs to be imported with `use`.

**Comment by fabpot** [comment]
ok, not worth it then.

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php [hunk 1]
```diff
@@ -167,4 +194,17 @@ public function endTest(\PHPUnit_Framework_Test $test, $time)
...
     }
+
+    public function handleError($type, $msg, $file, $line, $context)
+    {
+        if (E_USER_DEPRECATED !== $type && E_DEPRECATED !== $type) {
+            $h = $this->previousErrorHandler;
+
+            return $h ? $h($type, $msg, $file, $line, $context) : false;
+        }
+        if (error_reporting()) {
```
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: `0 !== error_reporting()` as we are dealing with integers here?

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 2]
```diff
@@ -566,19 +565,17 @@ public function testSerializerCacheDisabled()
 
     /**
      * @group legacy
-     * @requires function Symfony\Bridge\PhpUnit\ErrorAssert::assertDeprecationsAreTriggered
+     * @expectedDeprecation The "framework.serializer.cache" option is deprecated %s.
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: Before, this test wouldn't have been executed, when the PhpUnitBridge was too old. Is there any risk in always running tests now (except that the expected report may be wrong)?

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php [hunk 3]
```diff
@@ -151,13 +154,37 @@ public function startTest(\PHPUnit_Framework_Test $test)
...
         }
     }
 
     public function endTest(\PHPUnit_Framework_Test $test, $time)
     {
+        if ($this->expectedDeprecations) {
+            restore_error_handler();
+            try {
+                $prefix = "@expectedDeprecation:\n  ";
+                $test->assertStringMatchesFormat($prefix.implode("\n  ", $this->expectedDeprecations), $prefix.implode("\n  ", $this->gatheredDeprecations));
```
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: This now differs in one detail from the current implementation: Previously, the order in which you declared multiple expected deprecations did not matter. Now you need to know in which order they will be triggered.

File: src/Symfony/Bundle/FrameworkBundle/Tests/Templating/TemplateNameParserTest.php [hunk 4]
```diff
@@ -86,17 +85,15 @@ public function testParseValidNameWithNotFoundBundle()
     /**
      * @group legacy
      * @dataProvider provideAbsolutePaths
-     * @requires function Symfony\Bridge\PhpUnit\ErrorAssert::assertDeprecationsAreTriggered
+     * @expectedDeprecation Absolute template path support is deprecated since Symfony 3.1 and will be removed in 4.0.
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Templating/TemplateNameParserTest.php]: We will need to update tests in older branches too (see https://github.com/symfony/symfony/blob/3.1/src/Symfony/Bundle/FrameworkBundle/Tests/Templating/TemplateNameParserTest.php#L89, for an example). Updating them should be done in separate PRs (except for deprecations introduced in 3.2).

**Comment by xabbuh** [comment]
Except for the few comments I much better like this implementation. 👍

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: I disagree here: this practice makes the code *more* fragile, because it needs one more assumption to be correct. When I read this kind of code, I have to verify how `null`, `false` and `array()` are handled. Resorting to behavioral statements saying "there is no pb *here* because whatever", is the *hint* that this is bad practice.

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListener.php]: Fine with me :)

File: src/Symfony/Bundle/FrameworkBundle/Tests/Templating/TemplateNameParserTest.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Templating/TemplateNameParserTest.php]: Sure, and all at once (there is no way to make things always green step by step)

**Comment by nicolas-grekas** [comment]
Doc PR sent: https://github.com/symfony/symfony-docs/pull/7074

**Comment by nicolas-grekas** [comment]
This is now mergeable asis: ErrorAssert is not removed in this PR, and still used for deprecations introduced in 3.1. Those are removed in #20268.

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
